tests/io/test_write_modes.py

```diff
@@ -122,95 +178,121 @@ def test_write_modes_local_empty_data(tmp_path, write_mode, format):
+def test_write_modes_local_empty_data(tmp_path, write_mode, format):
+    _run_write_modes_empty_test(
+        path=str(tmp_path),
+        write_mode=write_mode,
+        format=format,
+        io_config=None,
     )
-    if not fs.exists(BUCKET):
-        fs.mkdir(BUCKET)
-    yield BUCKET
 
 
 @pytest.mark.integration()
 @pytest.mark.parametrize("write_mode", ["append", "overwrite"])
 @pytest.mark.parametrize("format", ["csv", "parquet"])
-@pytest.mark.parametrize("num_partitions", [1, 2])
-@pytest.mark.parametrize("partition_cols", [None, ["a"]])
-def test_write_modes_s3_minio(
+def test_write_modes_s3_minio_empty_data(
     minio_io_config,
     bucket,
     write_mode,
     format,
-    num_partitions,
-    partition_cols,
 ):
-    path = f"s3://{bucket}/{uuid.uuid4()!s}"
+    _run_write_modes_empty_test(
+        path=f"s3://{bucket}/{uuid.uuid4()!s}",
+        write_mode=write_mode,
+        format=format,
+        io_config=minio_io_config,
+    )
+
+
+OVERWRITE_PARTITION_TEST_CASES = [
+    pytest.param(
+        {
+            "a": ["a", "a", "b", "b"],
+            "b": [5, 6, 7, 8],
+        },
+        {
+            "a": ["a", "a", "b", "b"],
+            "b": [5, 6, 7, 8],
+        },
+        id="overwrite-all",
+    ),
+    pytest.param(
+        {
+            "a": ["a", "a"],
+            "b": [5, 6],
+        },
+        {
+            "a": ["a", "a", "b", "b"],
+            "b": [5, 6, 3, 4],
+        },
+        id="overwrite-some",
+    ),
+    pytest.param(
+        {
+            "a": ["b", "b", "c", "c"],
+            "b": [9, 10, 11, 12],
+        },
+        {
+            "a": ["a", "a", "b", "b", "c", "c"],
+            "b": [1, 2, 9, 10, 11, 12],
+        },
+        id="overwrite-and-append",
+    ),
+]
+
+
+def _run_overwrite_partitions_test(
+    path,
+    format,
+    new_data,
+    expected_read_back,
+    io_config,
+):
     existing_data = {"a": ["a", "a", "b", "b"], "b": [1, 2, 3, 4]}
-    new_data = {
-        "a": ["a", "a", "b", "b"],
-        "b": [5, 6, 7, 8],
-    }
 
     read_back = arrange_write_mode_test(
-        daft.from_pydict(existing_data).into_partitions(num_partitions),
-        daft.from_pydict(new_data).into_partitions(num_partitions),
+        daft.from_pydict(existing_data),
+        daft.from_pydict(new_data),
         path,
         format,
-        write_mode,
-        partition_cols,
-        minio_io_config,
+        "overwrite-partitions",
+        ["a"],
+        io_config,
     )
 
     # Check the data
-    if write_mode == "append":
-        assert read_back["a"] == ["a"] * 4 + ["b"] * 4
-        assert read_back["b"] == [1, 2, 5, 6, 3, 4, 7, 8]
-    elif write_mode == "overwrite":
-        assert read_back["a"] == ["a", "a", "b", "b"]
-        assert read_back["b"] == [5, 6, 7, 8]
-    else:
-        raise ValueError(f"Unsupported write_mode: {write_mode}")
+    for col in expected_read_back:
+        assert read_back[col] == expected_read_back[col]
```

Comment:
Hmm I guess we're checking that the data is correct, but maybe we should also check that only the expected files were modified?

One way to do this is by checking the mtimes of the files read. Since mtime typically has a resolution of 1s, we can also sleep for 1s before performing the write.